pyterrier_dr/bge_m3.py

```diff
@@ -0,0 +1,77 @@
+from pyterrier.transformer import Transformer
+from FlagEmbedding import BGEM3FlagModel
+from tqdm import tqdm
+import pyterrier as pt
+import pandas as pd
+import numpy as np
+import torch
+from .biencoder import BiEncoder
+
+
+class BGEM3Factory(BiEncoder):
+    def __init__(self, model_name='BAAI/bge-m3', batch_size=32, max_length=8192, text_field='text', verbose=False, device=None, use_fp16=False):
+        super().__init__(batch_size, text_field, verbose)
+        self.model_name = model_name
+        self.use_fp16 = use_fp16
+        self.max_length = max_length
+        if device is None:
+            device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        self.device = torch.device(device)
+
+        self.model = BGEM3FlagModel(self.model_name, use_fp16=self.use_fp16, device=self.device)
+
+
+    def __repr__(self):
+        return f'BGEM3Factory({repr(self.model_name)})'
+    
+    def encoder(self, verbose=None, batch_size=None, max_length=None) -> pt.Transformer:
+        '''
+        Encoding using Single-Vector Dense Embeddings
+        '''
+        return BGEM3Encoder(self, verbose=verbose, batch_size=batch_size, max_length=max_length)
+
+class BGEM3Encoder(pt.Transformer):
+    def __init__(self, bge_factory: BGEM3Factory, verbose=None, batch_size=None, max_length=None):
+        self.bge_factory = bge_factory
+        self.verbose = verbose if verbose is not None else bge_factory.verbose
+        self.batch_size = batch_size if batch_size is not None else bge_factory.batch_size
+        self.max_length = max_length if max_length is not None else bge_factory.max_length
+
+    def encode(self, texts):
+        return self.bge_factory.model.encode(list(texts), batch_size=self.batch_size, max_length=self.max_length,
+                             return_dense=True, return_sparse=True, return_colbert_vecs=True)
+
+    def transform(self, inp: pd.DataFrame) -> pd.DataFrame:
+        if 'query' in inp.columns:
+            # assert all(c in inp.columns for c in ['query'])
+            it = inp['query'].values
+            it, inv = np.unique(it, return_inverse=True)
+            if self.verbose:
+                it = pt.tqdm(it, desc='Encoding Single Vector Embeddings', unit='query')
+            bgem3_results = self.encode(it)
+            query_vec = [bgem3_results['dense_vecs'][i] for i in inv]
+            query_sparse = [bgem3_results['lexical_weights'][i] for i in inv]
+            query_multivecs = [bgem3_results['colbert_vecs'][i] for i in inv]
+            # add to the input dataframe
+            inp = inp.assign(query_vec=query_vec,
```

Comment:
I think it'd be nice to have a way to configure which columns are included. The main reason is that some fields (especially multivec) can get pretty big, and if they're not needed, it would be nice not to include them.

The encoder class could take in flags for each type? Then the main class could have several encoder factories:
 - `.query_multi_encoder()` and `.doc_multi_encoder()` (with flags)
 - `.query_encoder()` and `.doc_encoder()` (single-vec, matching the naming convention in this package)
 - `.query_multivec_encoder()` and `.doc_multivec_encoder()`
 - `.query_sparse_encoder()` and `.doc_sparse_encoder()`